src/tsd.c

```diff
@@ -23,7 +23,11 @@ pthread_key_t tsd_tsd;
 bool tsd_booted = false;
 #elif (defined(_WIN32))
 DWORD tsd_tsd;
-tsd_wrapper_t tsd_boot_wrapper = {false, TSD_INITIALIZER};
+#if defined(JEMALLOC_LEGACY_WINDOWS_SUPPORT) || !defined(_MSC_VER)
+tsd_wrapper_t tsd_boot_wrapper = {TSD_INITIALIZER, false};
+#else
+JEMALLOC_TSD_TYPE_ATTR(tsd_wrapper_t) tsd_wrapper_tls = { TSD_INITIALIZER, false };
```

Comment:
I'm a bit lost here. How does the underlying storage work here? Does this already reserve the space for each thread? Or it still needs to do TlsGet / SetValue? In tsd_boot0 there is TlsAlloc, which to my understanding isn't needed for __declspec( thread )?

Reply:
The way __declspec(thread) works is that variables marked with __declspec(thread) will have a section in the PE file (exe/dll), see https://learn.microsoft.com/en-us/windows/win32/debug/pe-format. The Windows loader looks at the PE file and dynamically allocates a thread-specific region for each newly created thread and copies the initial data from the PE file, and the tls region is reachable from TEB (see https://learn.microsoft.com/en-us/windows/win32/api/winternl/ns-winternl-teb). The MSVC simply emits code that traverses TEB to access the tls block. Simply put, although not 100% accurate, you may consider the compiler inlines TlsGetValue when accessing __declspec(thread). This removes the need to manually manage tls memory allocation/deallocation, and eliminates TlsGetValue calls.

Reply:
I see your point, thanks. Yes, TlsAlloc is not needed here, sorry I missed it.. To be clear, TlsAlloc doesn't allocate memory for thread-local objects. Instead, it merely allocates an `index` that every thread can use to successfully call TlsSetValue to index a tls object. TlsAlloc is similar to pthread_setspecific, except that it doesn't re-arm the thread clean up routine. That's why on windows, we use a wrapper object with an additional `initialized` field, which indicates whether we need to call thread cleanup routine again.